libsolidity/ExpressionCompiler.cpp

```diff
@@ -823,26 +823,85 @@ unsigned ExpressionCompiler::appendArgumentCopyToMemory(TypePointers const& _typ
 	return length;
 }
 
-unsigned ExpressionCompiler::appendExpressionCopyToMemory(Type const& _expectedType,
-														  Expression const& _expression, unsigned _memoryOffset)
+unsigned ExpressionCompiler::appendTypeCopyToMemory(Type const& _expectedType, TypePointer const& _type,
+													Location const& _location, unsigned _memoryOffset)
 {
-	_expression.accept(*this);
-	appendTypeConversion(*_expression.getType(), _expectedType, true);
+	appendTypeConversion(*_type, _expectedType, true);
 	unsigned const c_numBytes = CompilerUtils::getPaddedSize(_expectedType.getCalldataEncodedSize());
 	if (c_numBytes == 0 || c_numBytes > 32)
 		BOOST_THROW_EXCEPTION(CompilerError()
-							  << errinfo_sourceLocation(_expression.getLocation())
+							  << errinfo_sourceLocation(_location)
 							  << errinfo_comment("Type " + _expectedType.toString() + " not yet supported."));
 	bool const c_leftAligned = _expectedType.getCategory() == Type::Category::STRING;
 	bool const c_padToWords = true;
 	return CompilerUtils(m_context).storeInMemory(_memoryOffset, c_numBytes, c_leftAligned, c_padToWords);
 }
 
+unsigned ExpressionCompiler::appendExpressionCopyToMemory(Type const& _expectedType,
+														  Expression const& _expression,
+														  unsigned _memoryOffset)
+{
+	_expression.accept(*this);
+	return appendTypeCopyToMemory(_expectedType, _expression.getType(), _expression.getLocation(), _memoryOffset);
+}
+
 void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const& _varDecl)
 {
-	m_currentLValue.fromStateVariable(_varDecl, _varDecl.getType());
-	solAssert(m_currentLValue.isInStorage(), "");
-	m_currentLValue.retrieveValue(_varDecl.getType(), Location(), true);
+	auto mappingType = dynamic_cast<const MappingType*>(_varDecl.getType().get());
+	unsigned sizeOnStack;
+	if (mappingType != nullptr)
+	{
+		// this copies from Compiler::visit(FunctionDefinition..) for argument reading
+		unsigned parameterSize = mappingType->getKeyType()->getSizeOnStack();
+		m_context.adjustStackOffset(parameterSize);
+		m_context.addVariable(_varDecl, parameterSize);
+		// this copies from ExpressionCompiler::visit(IndexAccess .. ) for mapping access
+		TypePointer const& keyType = mappingType->getKeyType();
+		unsigned length = appendTypeCopyToMemory(*keyType, mappingType->getValueType(), Location());
+		solAssert(length == 32, "Mapping key has to take 32 bytes in memory (for now).");
+		// @todo move this once we actually use memory
+		length += CompilerUtils(m_context).storeInMemory(length);
```

Comment:
You need to retrieve the position of the mapping `getStorageLocationOfVariable` to the stack, too. In `visit(IndexAccess` this is done by `_indexAccess.getBaseExpression().accept(*this);`.